File: benchmarks/data_wrangling/bench.py
"""bench.py -- Table/DataFrame operations benchmark, Python (pandas) side.

Same dataset (dataset.csv, built once by generate_data.py) and same six
steps as bench.qu/bench.m: load, filter, groupby+agg, sort, descriptive
stats, correlation matrix. Prints both timings and the actual computed
values so results can be cross-checked against the other two languages,
not just timed.

Run: python benchmarks/data_wrangling/bench.py
"""
import os
import time

import numpy as np
import pandas as pd

HERE = os.path.dirname(os.path.abspath(__file__))
CSV_PATH = os.path.join(HERE, "dataset.csv")

NUM_COLS = ["x1", "x2", "x3", "x4"]

# --- 1. load --------------------------------------------------------------
t0 = time.perf_counter()
df = pd.read_csv(CSV_PATH)
t_load = time.perf_counter() - t0
print(f"load          : {t_load:.4f} s  ({len(df)} rows, {df.shape[1]} cols)")

# --- 2. filter: x1 > 0 -----------------------------------------------------
t0 = time.perf_counter()
filtered = df[df["x1"] > 0]
t_filter = time.perf_counter() - t0
print(f"filter        : {t_filter:.4f} s  ({len(filtered)} rows kept)")

# --- 3. groupby(category) + agg(mean/sum/count of x2) ----------------------
t0 = time.perf_counter()
grouped = df.groupby("category")["x2"].agg(["mean", "sum", "count"])
t_group = time.perf_counter() - t0
grouped_sorted = grouped.sort_index()
print(f"groupby_agg   : {t_group:.4f} s  ({len(grouped)} groups)")

# --- 4. sort by x3 -----------------------------------------------------------
t0 = time.perf_counter()
sorted_df = df.sort_values("x3")
t_sort = time.perf_counter() - t0
print(f"sort_by       : {t_sort:.4f} s")

# --- 5. descriptive stats: describe() on numeric columns --------------------
t0 = time.perf_counter()
desc = df[NUM_COLS].describe()
t_describe = time.perf_counter() - t0
print(f"describe      : {t_describe:.4f} s")

# --- 6. correlation matrix ---------------------------------------------------
t0 = time.perf_counter()
corr = df[NUM_COLS].corr()
t_corr = time.perf_counter() - t0
print(f"corr_matrix   : {t_corr:.4f} s")

total = t_load + t_filter + t_group + t_sort + t_describe + t_corr
print(f"total         : {total:.4f} s")

# --- correctness cross-check printout ---------------------------------------
print()
print("=== correctness cross-check values ===")
print(f"total_rows          = {len(df)}")
print(f"filtered_rows(x1>0) = {len(filtered)}")
print("group_counts (sorted by category):")
for cat, row in grouped_sorted.iterrows():
    print(f"  {cat}: mean={row['mean']:.6f} sum={row['sum']:.6f} count={int(row['count'])}")
print(f"sorted_first_x3     = {sorted_df['x3'].iloc[0]:.6f}")
print(f"sorted_last_x3      = {sorted_df['x3'].iloc[-1]:.6f}")
print(f"sorted_is_monotonic = {bool(sorted_df['x3'].is_monotonic_increasing)}")
print("describe():")
print(desc.to_string())
print("corr matrix:")
print(corr.to_string())

File: benchmarks/data_wrangling/bench.m
% bench.m -- Table/DataFrame operations benchmark, MATLAB side.
%
% Same dataset (dataset.csv, built once by generate_data.py) and same six
% steps as bench.qu/bench.py: load, filter, groupby+agg, sort, descriptive
% stats, correlation matrix -- using MATLAB's own `table`/`readtable`,
% logical indexing, `groupsummary`, `sortrows`, and `summary`/`corrcoef`.
%
% Run: matlab -batch "run('benchmarks/data_wrangling/bench.m')"

csv_path = 'benchmarks/data_wrangling/dataset.csv';
num_cols = {'x1', 'x2', 'x3', 'x4'};

% --- 1. load -----------------------------------------------------------------
tic;
T = readtable(csv_path);
t_load = toc;
fprintf('load          : %.4f s  (%d rows, %d cols)\n', t_load, height(T), width(T));

% --- 2. filter: x1 > 0 --------------------------------------------------------
tic;
filtered = T(T.x1 > 0, :);
t_filter = toc;
fprintf('filter        : %.4f s  (%d rows kept)\n', t_filter, height(filtered));

% --- 3. groupby(category) + agg(mean/sum/count of x2) -------------------------
% NOTE: `groupsummary`'s own 'nnz' method (nonzero count) is NOT the same
% as a row count -- it silently undercounts if any value in the group
% happens to equal exactly 0 (found the hard way: one x2 value in this
% dataset rounds to 0.000000 at 6 decimal places, so an earlier version of
% this script using 'nnz' reported 999,999 total rows instead of 1,000,000
% when cross-checked against pandas' `.agg(["count"])`). `groupsummary`
% always returns a `GroupCount` column (true per-group row count,
% independent of any requested aggregation methods) -- use that instead.
tic;
grouped = groupsummary(T, 'category', {'mean', 'sum'}, 'x2');
t_group = toc;
grouped = sortrows(grouped, 'category');
fprintf('groupby_agg   : %.4f s  (%d groups)\n', t_group, height(grouped));

% --- 4. sort by x3 -------------------------------------------------------------
tic;
sorted_T = sortrows(T, 'x3');
t_sort = toc;
fprintf('sort_by       : %.4f s\n', t_sort);

% --- 5. descriptive stats: count/mean/std/min/25/50/75/max on numeric cols ----
tic;
X = T{:, num_cols};
n = size(X, 1);
stat_count = repmat(n, 1, 4);
stat_mean = mean(X, 1);
stat_std = std(X, 0, 1);
stat_min = min(X, [], 1);
stat_p25 = prctile(X, 25, 1);
stat_p50 = prctile(X, 50, 1);
stat_p75 = prctile(X, 75, 1);
stat_max = max(X, [], 1);
t_describe = toc;
fprintf('describe      : %.4f s\n', t_describe);

% --- 6. correlation matrix ---------------------------------------------------
tic;
C = corrcoef(X);
t_corr = toc;
fprintf('corr_matrix   : %.4f s\n', t_corr);

total = t_load + t_filter + t_group + t_sort + t_describe + t_corr;
fprintf('total         : %.4f s\n', total);

% --- correctness cross-check printout ---------------------------------------
fprintf('\n=== correctness cross-check values ===\n');
fprintf('total_rows          = %d\n', height(T));
fprintf('filtered_rows(x1>0) = %d\n', height(filtered));
fprintf('group means (grp_0..grp_7, alphabetical): ');
fprintf('%.6f ', grouped.mean_x2);
fprintf('\n');
fprintf('group sums  (grp_0..grp_7, alphabetical): ');
fprintf('%.6f ', grouped.sum_x2);
fprintf('\n');
fprintf('group counts(grp_0..grp_7, alphabetical): ');
fprintf('%d ', grouped.GroupCount);
fprintf('\n');
fprintf('sorted_first_x3     = %.6f\n', sorted_T.x3(1));
fprintf('sorted_last_x3      = %.6f\n', sorted_T.x3(end));
fprintf('describe() stat order: count, mean, std, min, 25%%, 50%%, 75%%, max\n');
fprintf('describe().x1 = %.6f %.6f %.6f %.6f %.6f %.6f %.6f %.6f\n', stat_count(1), stat_mean(1), stat_std(1), stat_min(1), stat_p25(1), stat_p50(1), stat_p75(1), stat_max(1));
fprintf('describe().x2 = %.6f %.6f %.6f %.6f %.6f %.6f %.6f %.6f\n', stat_count(2), stat_mean(2), stat_std(2), stat_min(2), stat_p25(2), stat_p50(2), stat_p75(2), stat_max(2));
fprintf('describe().x3 = %.6f %.6f %.6f %.6f %.6f %.6f %.6f %.6f\n', stat_count(3), stat_mean(3), stat_std(3), stat_min(3), stat_p25(3), stat_p50(3), stat_p75(3), stat_max(3));
fprintf('describe().x4 = %.6f %.6f %.6f %.6f %.6f %.6f %.6f %.6f\n', stat_count(4), stat_mean(4), stat_std(4), stat_min(4), stat_p25(4), stat_p50(4), stat_p75(4), stat_max(4));
fprintf('corr matrix:\n');
disp(C);
